README
NAME
    Linux::SysInfo - Perl interface to the sysinfo(2) Linux system call.

VERSION
    Version 0.13

SYNOPSIS
        use Linux::SysInfo qw/sysinfo/;

        my $si = sysinfo;
        print "$_: $si->{$_}\n" for keys %$si;

DESCRIPTION
    This module is a wrapper around the sysinfo(2) Linux system call. It
    gives information about the current uptime, load average, memory usage
    and processes running. Other systems have also this system call (e.g.
    Solaris), but in most cases the returned information is different.

CONSTANTS
  "LS_HAS_EXTENDED"
    This constant is set to 1 if your kernel supports the three extended
    fields "totalhigh", "freehigh" and "mem_unit" ; and to 0 otherwise.

FUNCTIONS
  "sysinfo"
    This function takes no argument. It returns "undef" on failure or a hash
    reference whose keys are the members name of the "struct sysinfo" on
    success :

    *   "uptime"

        Seconds elapsed since the system booted.

    *   "load1", "load5", "load15"

        1, 5 and 15 minutes load average.

    *   "totalram"

        Total usable main memory size.

    *   "freeram"

        Available memory size.

    *   "sharedram"

        Amount of shared memory.

    *   "bufferram"

        Memory used by buffers.

    *   "totalswap"

        Total swap space size.

    *   "freeswap"

        Swap space still available.

    *   "procs"

        Number of current processes.

    Prior to Linux 2.3.23 on i386 and 2.3.48 on all other architectures, the
    memory sizes were given in bytes. Since then, the following members are
    also available and all the memory sizes are given as multiples of
    "mem_unit" bytes :

    *   "totalhigh"

        Total high memory size.

    *   "freehigh"

        Available high memory size.

    *   "mem_unit"

        Memory unit size in bytes.

EXPORT
    The only function of this module, "sysinfo", and the constant
    "LS_HAS_EXTENDED" are only exported on request. Functions are also
    exported by the ":funcs" tag, and constants by ":consts".

BINARY COMPATIBILITY
    If you upgrade your kernel to a greater version than 2.3.23 on i386 or
    2.3.48 on any other platform, you will need to rebuild the module to
    access the extended fields.

    Moreover, since the perl hash function has changed after the 5.6
    version, you will also need to recompile the module if you upgrade your
    perl from a version earlier than 5.6.

DEPENDENCIES
    perl 5.6.

SEE ALSO
    The sysinfo(2) man page.

    Sys::Info : Gather information about your system.

    Sys::CpuLoad : Try several different methods to retrieve the load
    average.

    BSD::getloadavg : Wrapper to the getloadavg(3) BSD system call.

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

    You can contact me by mail or on "irc.perl.org" (vincent).

BUGS
    Please report any bugs or feature requests to "bug-linux-sysinfo at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Linux-SysInfo>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Linux::SysInfo

    Tests code coverage report is available at
    <http://www.profvince.com/perl/cover/Linux-SysInfo>.

COPYRIGHT & LICENSE
    Copyright 2007,2008,2009,2010 Vincent Pit, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Linux-SysInfo

0.13    2010-03-05 22:50 UTC
        + Fix : t/30-threads.t will not be correctly skipped for perls with
                ithreads but without threads.pm. Thanks David Cantrell for
                providing access to a machine with such a perl.

0.12    2010-01-07 17:50 UTC
        + Fix : Work around Kwalitee test misfailures.
        + Tst : Test thread safety.

0.11    2009-03-10 16:25 UTC
        + Chg : Perl 5.6 is required.
        + Doc : Some formatting cleanup.
        + Tst : Check that the returned values look like numbers.
        + Upd : META.yml spec updated to 1.4.

0.10    2008-03-10 15:40 UTC
        + Doc : The usual set of POD typos.
        + Fix : Ensure that we're on linux with $^O, as some smokers don't seem
                to apply anymore the rule of "if it's in the Linux:: namespace,
                then errors should be reported as FAILures only if we actually
                are on linux".
        + Tst : Tiny refinements.

0.09    2008-03-09 13:00 UTC
        + Add : ':funcs' and ':consts' export tags.
        + Doc : Copyright update.
        + Fix : Correct dependencies listing in META.yml.
        + Tst : Author tests overhaul.
        + Tst : t/95-portability-files.t.

0.08    2007-11-19 09:25 UTC
        + Fix : 5.10.0_RC1 compatibility fix.

0.07    2007-08-28 12:30 UTC
        + Chg : s/require (XSLoader)/use $1/.
        + Fix : Tests are now strict.

0.06    2007-08-14 18:40 UTC
        + Fix : Building for perl-5.6.x and perl-5.9.{3,4}.

0.05    2007-08-13 11:10 UTC
        + Add : Kwalitee test.
	+ Chg : Makefile.PL overhaul.
        + Chg : Hash keys are now shared. This new implementation doesn't like
                5.9.3, but seems to be somehow ok with 5.9.5. If you're using
                5.9.3, either upgrade or keep with the precedent module version.

0.04	2007-05-03 09:35 UTC
	+ Add : The samples/sysinfo.pl sample script.
	+ Chg : Declare LS_HAS_EXTENDED with newCONSTSUB.
	+ Fix : Missing LICENSE in Makefile.PL
	+ Fix : Skipping extended tests will now display the good reason.

0.03	2007-04-18 10:25 UTC
	+ Fix : Load average raw values have to be shifted on a platform
	        dependant way. They are now correctly reported.
	+ Fix : Missing META.yml
	+ Doc : Information about binary compatibility.

0.01    2007-04-11 12:45 UTC
        * First version, released on an unsuspecting world.



Makefile.PL
use 5.006;

use strict;
use warnings;
use ExtUtils::MakeMaker;

die 'OS unsupported' unless $^O && $^O eq 'linux';

my $dist = 'Linux-SysInfo';

(my $name = $dist) =~ s{-}{::}g;

(my $file = $dist) =~ s{-}{/}g;
$file = "lib/$file.pm";

my %PREREQ_PM = (
 'Exporter' => 0,
 'XSLoader' => 0,
 'base'     => 0,
);

my %META = (
 configure_requires => {
  'ExtUtils::MakeMaker' => 0,
 },
 build_requires => {
  'ExtUtils::MakeMaker' => 0,
  'Test::More'          => 0,
  %PREREQ_PM,
 },
 dynamic_config => 1,
 resources => {
  bugtracker => "http://rt.cpan.org/NoAuth/ReportBug.html?Queue=$dist",
  homepage   => "http://search.cpan.org/dist/$dist/",
  license    => 'http://dev.perl.org/licenses/',
  repository => "http://git.profvince.com/?p=perl%2Fmodules%2F$dist.git",
 },
);

WriteMakefile(
 NAME             => $name,
 AUTHOR           => 'Vincent Pit <perl@profvince.com>',
 LICENSE          => 'perl',
 VERSION_FROM     => $file,
 ABSTRACT_FROM    => $file,
 PL_FILES         => {},
 PREREQ_PM        => \%PREREQ_PM,
 MIN_PERL_VERSION => 5.006,
 META_MERGE       => \%META,
 dist             => {
  PREOP    => "pod2text $file > \$(DISTVNAME)/README",
  COMPRESS => 'gzip -9f', SUFFIX => 'gz',
 },
 clean            => {
  FILES => "$dist-* *.gcov *.gcda *.gcno cover_db Debian_CPANTS.txt"
 },
);


lib/Linux/SysInfo.pm
package Linux::SysInfo;

use strict;
use warnings;

=head1 NAME

Linux::SysInfo - Perl interface to the sysinfo(2) Linux system call.

=head1 VERSION

Version 0.13

=cut

our $VERSION;
BEGIN {
 $VERSION = '0.13';
}

=head1 SYNOPSIS

    use Linux::SysInfo qw/sysinfo/;

    my $si = sysinfo;
    print "$_: $si->{$_}\n" for keys %$si;

=head1 DESCRIPTION

This module is a wrapper around the C<sysinfo(2)> Linux system call.
It gives information about the current uptime, load average, memory usage and processes running.
Other systems have also this system call (e.g. Solaris), but in most cases the returned information is different.

=head1 CONSTANTS

=head2 C<LS_HAS_EXTENDED>

This constant is set to C<1> if your kernel supports the three extended fields C<totalhigh>, C<freehigh> and C<mem_unit> ; and to C<0> otherwise.

=head1 FUNCTIONS

=cut

BEGIN {
 require XSLoader;
 XSLoader::load(__PACKAGE__, $VERSION);
}

=head2 C<sysinfo>

This function takes no argument.
It returns C<undef> on failure or a hash reference whose keys are the members name of the C<struct sysinfo> on success :

=over 4

=item *

C<uptime>

Seconds elapsed since the system booted.

=item *

C<load1>, C<load5>, C<load15>

1, 5 and 15 minutes load average.

=item *

C<totalram>

Total usable main memory size.

=item *

C<freeram>

Available memory size.

=item *

C<sharedram>

Amount of shared memory.

=item *

C<bufferram>

Memory used by buffers.

=item *

C<totalswap>

Total swap space size.

=item *

C<freeswap>

Swap space still available.

=item *

C<procs>

Number of current processes.

=back

Prior to Linux 2.3.23 on i386 and 2.3.48 on all other architectures, the memory sizes were given in bytes.
Since then, the following members are also available and all the memory sizes are given as multiples of C<mem_unit> bytes :

=over 4

=item *

C<totalhigh>

Total high memory size.

=item *

C<freehigh>

Available high memory size.

=item *

C<mem_unit>

Memory unit size in bytes.

=back

=head1 EXPORT

The only function of this module, C<sysinfo>, and the constant C<LS_HAS_EXTENDED> are only exported on request.
Functions are also exported by the C<:funcs> tag, and constants by C<:consts>.

=cut

use base qw/Exporter/;

our @EXPORT         = ();
our %EXPORT_TAGS    = (
 'funcs'  => [ qw/sysinfo/ ],
 'consts' => [ qw/LS_HAS_EXTENDED/ ]
);
our @EXPORT_OK      = map { @$_ } values %EXPORT_TAGS;
$EXPORT_TAGS{'all'} = [ @EXPORT_OK ];

=head1 BINARY COMPATIBILITY

If you upgrade your kernel to a greater version than 2.3.23 on i386 or 2.3.48 on any other platform, you will need to rebuild the module to access the extended fields.

Moreover, since the perl hash function has changed after the 5.6 version, you will also need to recompile the module if you upgrade your perl from a version earlier than 5.6.

=head1 DEPENDENCIES

L<perl> 5.6.

=head1 SEE ALSO

The C<sysinfo(2)> man page.

L<Sys::Info> : Gather information about your system.

L<Sys::CpuLoad> : Try several different methods to retrieve the load average.

L<BSD::getloadavg> : Wrapper to the C<getloadavg(3)> BSD system call.

=head1 AUTHOR

Vincent Pit, C<< <perl at profvince.com> >>, L<http://www.profvince.com>.

You can contact me by mail or on C<irc.perl.org> (vincent).

=head1 BUGS

Please report any bugs or feature requests to C<bug-linux-sysinfo at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Linux-SysInfo>.
I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Linux::SysInfo

Tests code coverage report is available at L<http://www.profvince.com/perl/cover/Linux-SysInfo>.

=head1 COPYRIGHT & LICENSE

Copyright 2007,2008,2009,2010 Vincent Pit, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Linux::SysInfo
